src/swap/SwapReviewScreen.tsx

```diff
@@ -0,0 +1,374 @@
+
+const TAG = 'SWAP_REVIEW_SCREEN'
+
+export function SwapReviewScreen(props: Props) {
+  const { toToken, fromToken, swapAmount, updatedField } = props.route.params
+  const [loading, setLoading] = useState(false)
+  const [fetchSwapQuoteError, setFetchSwapQuoteError] = useState(false)
+  const [estimatedModalVisible, setEstimatedDialogVisible] = useState(false)
+  const [swapFeeModalVisible, setSwapFeeModalVisible] = useState(false)
+  const [swapInfo, setSwapInfo] = useState(null as any)
+  const coreTokens = useSelector(coreTokensSelector)
+  const walletAddress = useSelector(walletAddressSelector)
+
+  // Items set for remote config
+  const maxSlippagePercent = useSelector(maxSwapSlippagePercentageSelector)
+  const swapFeeEnabled = useSelector(swapFeeEnabledSelector)
+  const swapFeePercentage = useSelector(swapFeePercentageSelector)
+
+  // Remote configs converted to decimals strings
+  const maxSlippageDecimal = `${maxSlippagePercent / 100}`
+  const swapFeeDecimal = `${swapFeePercentage / 100}`
+
+  const { t } = useTranslation()
+  const dispatch = useDispatch()
+
+  // Token Symbols
+  const toTokenSymbol = coreTokens.find((token) => token.address === toToken)?.symbol
+  const fromTokenSymbol = coreTokens.find((token) => token.address === fromToken)?.symbol
+
+  useEffect(() => {
+    if (fetchSwapQuoteError) {
+      dispatch(showError(ErrorMessages.FETCH_SWAP_QUOTE_FAILED))
+    }
+  }, [fetchSwapQuoteError])
+
+  useEffect(() => {
+    ValoraAnalytics.track(SwapEvents.swap_review_screen_open, {
+      toToken,
+      fromToken,
+      buyAmount: swapAmount[Field.TO] ?? '',
+    })
+  }, [])
+
+  const getSwapInfo = async () => {
+    try {
+      setFetchSwapQuoteError(false)
+      setLoading(true)
+      const swapAmountInWei = multiplyByWei(swapAmount[updatedField]!)
+      const swapAmountParam = updatedField === Field.FROM ? 'sellAmount' : 'buyAmount'
+      const params = {
+        buyToken: toToken,
+        sellToken: fromToken,
+        [swapAmountParam]: swapAmountInWei.toString().split('.')[0],
+        slippagePercentage: maxSlippageDecimal,
+        userAddress: walletAddress,
+      }
+      const queryParams = new URLSearchParams({ ...(params as Record<string, string>) }).toString()
+      const requestUrl = `${networkConfig.approveSwapUrl}?${queryParams}`
+      const response = await fetch(requestUrl)
+      if (!response.ok) {
+        setFetchSwapQuoteError(true)
+        Logger.error(TAG, `Failure response fetching token Swap: ${response}`)
+        throw new Error(
+          `Failure response fetching token swap quote. ${response.status}  ${response.statusText}`
+        )
+      }
+      setSwapInfo(await response.json())
+    } catch (error) {
+      setFetchSwapQuoteError(true)
+      Logger.error(TAG, `Error fetching token swap quote: ${error}`)
+    } finally {
+      setLoading(false)
+    }
+  }
+
+  // Get swap info on load
+  useAsync(getSwapInfo, [])
```

Comment:
what do you think about the following idea:

- we use all the stuff that `useAsync` gives us out of the box, i.e. the loading / data / error return variables and the error handler
- we introduce a state variable (something like `shouldFetch`) and add this as a dependency for the `useAsync` hook. the refresh control can also toggle this variable to trigger a refetch.

the benefits of this approach would be less state variables (don't need `loading`, `fetchSwapQuoteError`, `swapInfo` anymore) and less manual coordination of what the loading / error state is. we could also display the error message in the error handler of `useAsync` and remove the `useEffect` currently used to achieve this. i guess the examples i'm thinking of are `queryHelper.ts` and `I18nGate.tsx`, where by using `useAsync` we no longer need to manually track the loading / data / error variables in our own state variables.

 happy to catch up about this i didn't explain this idea clearly

Reply:
Changed in https://github.com/valora-inc/wallet/pull/2767/commits/e286b05c2e78276f2027576d09ec0073d3b41ab4. I still had to use a fetchError state to disable the swap button if the fetch had completed with an error, and a swapInfo in state, but the loading state is now handled by the useAsync.